stock_picking_type_routing_operation/models/stock_move.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2019 Camptocamp (https://www.camptocamp.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from itertools import chain
+from odoo import models
+
+
+class StockMove(models.Model):
+    _inherit = 'stock.move'
+
+    def _action_assign(self):
+        super()._action_assign()
+        if not self.env.context.get('exclude_apply_routing_operation'):
+            moves = self._split_per_routing_operation()
+            moves._apply_move_location_routing_operation()
+
+    def _split_per_routing_operation(self):
+        move_to_assign_ids = set()
+        new_move_per_location = {}
+        for move in self:
+            if move.state not in ('assigned', 'partially_available'):
+                continue
+
+            # Group move lines per source location, some may need an additional
+            # operations while others not. Store the number of products to
+            # take from each location, so we'll be able to split the move
+            # if needed.
+            move_lines = {}
+            for move_line in move.move_line_ids:
+                location = move_line.location_id
+                move_lines[location] = sum(move_line.mapped('product_uom_qty'))
+
+            # We'll split the move to have one move per different location where
+            # we have to take products
+            routing_quantities = {}
+            for source_location, qty in move_lines.items():
+                routing_picking_type = \
+                    source_location._find_picking_type_for_routing_operation()
+                routing_quantities.setdefault(routing_picking_type, 0.0)
+                routing_quantities[routing_picking_type] += qty
+
+            if len(routing_quantities) == 1:
+                # The whole quantity can be taken from only one location (an
+                # empty routing picking type being equal to one location here),
+                # nothing to split.
+                continue
+
+            move._do_unreserve()
```

Comment:
action_assign can create package level, so you need to destroy it when you unreserve

```suggestion
            move._do_unreserve()
            move.package_level_id.unlink()
```